// adonis-autoswagger/dist/adonishelpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseBindingReference = exports.serializeV6Handler = exports.serializeV6Middleware = void 0;
function serializeV6Middleware(mw) {
    return [...mw.all()].reduce((result, one) => {
        if (typeof one === "function") {
            result.push(one.name || "closure");
            return result;
        }
        if ("name" in one && one.name) {
            result.push(one.name);
        }
        return result;
    }, []);
}
exports.serializeV6Middleware = serializeV6Middleware;
async function serializeV6Handler(handler) {
    /**
     * Value is a controller reference
     */
    if ("reference" in handler) {
        return {
            type: "controller",
            ...(await parseBindingReference(handler.reference)),
        };
    }
    /**
     * Value is an inline closure
     */
    return {
        type: "closure",
        name: handler.name || "closure",
    };
}
exports.serializeV6Handler = serializeV6Handler;
async function parseBindingReference(binding) {
    const parseImports = (await import("parse-imports")).default;
    /**
     * The binding reference is a magic string. It might not have method
     * name attached to it. Therefore we split the string and attempt
     * to find the method or use the default method name "handle".
     */
    if (typeof binding === "string") {
        const tokens = binding.split(".");
        if (tokens.length === 1) {
            return { moduleNameOrPath: binding, method: "handle" };
        }
        return { method: tokens.pop(), moduleNameOrPath: tokens.join(".") };
    }
    const [bindingReference, method] = binding;
    /**
     * Parsing the binding reference for dynamic imports and using its
     * import value.
     */
    const imports = [...(await parseImports(bindingReference.toString()))];
    const importedModule = imports.find(($import) => $import.isDynamicImport && $import.moduleSpecifier.value);
    if (importedModule) {
        return {
            moduleNameOrPath: importedModule.moduleSpecifier.value,
            method: method || "handle",
        };
    }
    /**
     * Otherwise using the name of the binding reference.
     */
    return {
        moduleNameOrPath: bindingReference.name,
        method: method || "handle",
    };
}
exports.parseBindingReference = parseBindingReference;
