// @poppinss/intl-formatter/build/index.js
var __defProp = Object.defineProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/formatters.ts
var formatters_exports = {};
__export(formatters_exports, {
  date: () => date,
  displayNames: () => displayNames,
  list: () => list,
  number: () => number,
  plural: () => plural,
  relative: () => relative
});
import moize from "moize";
var number = moize.default(
  (...args) => {
    return new Intl.NumberFormat(...args);
  },
  { isDeepEqual: true }
);
var date = moize.default(
  (...args) => {
    return new Intl.DateTimeFormat(...args);
  },
  { isDeepEqual: true }
);
var plural = moize.default(
  (...args) => {
    return new Intl.PluralRules(...args);
  },
  { isDeepEqual: true }
);
var relative = moize.default(
  (...args) => {
    return new Intl.RelativeTimeFormat(...args);
  },
  { isDeepEqual: true }
);
var list = moize.default(
  (...args) => {
    return new Intl.ListFormat(...args);
  },
  { isDeepEqual: true }
);
var displayNames = moize.default(
  (...args) => {
    return new Intl.DisplayNames(...args);
  },
  { isDeepEqual: true }
);
export {
  formatters_exports as default
};
